ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -122,15 +123,16 @@ public ReplicaThread(String threadName, FindTokenHelper findTokenHelper, Cluster
       ReplicaSyncUpManager replicaSyncUpManager) {
     this(threadName, findTokenHelper, clusterMap, correlationIdGenerator, dataNodeId, connectionPool, replicationConfig,
         replicationMetrics, notification, storeKeyConverter, transformer, metricRegistry, replicatingOverSsl,
-        datacenterName, responseHandler, time, replicaSyncUpManager, null);
+        datacenterName, responseHandler, time, replicaSyncUpManager, null, null);
   }
 
   public ReplicaThread(String threadName, FindTokenHelper findTokenHelper, ClusterMap clusterMap,
       AtomicInteger correlationIdGenerator, DataNodeId dataNodeId, ConnectionPool connectionPool,
       ReplicationConfig replicationConfig, ReplicationMetrics replicationMetrics, NotificationSystem notification,
       StoreKeyConverter storeKeyConverter, Transformer transformer, MetricRegistry metricRegistry,
       boolean replicatingOverSsl, String datacenterName, ResponseHandler responseHandler, Time time,
-      ReplicaSyncUpManager replicaSyncUpManager, PartitionLeaderInfo partitionLeaderInfo) {
+      ReplicaSyncUpManager replicaSyncUpManager, PartitionLeaderInfo partitionLeaderInfo,
+      ReplicationEngine replicationEngine) {
```

Comment:
We probably don't need to pass in `ReplicationEngine`, instead we can pass in something like `leaderBasedReplicationTracker` or `leaderBasedReplicationCoordinator` or `leaderBasedReplicationAdmin` that helps to coordinate/update info across replication threads. The component is instantiated only when model == `ReplicationModelType.LEADER_BASED`. Thus, we can check if this component is null and don't have to check many other things like line350~351

Reply:
If we decide to pass in a component rather than `ReplicationEngine`, we can even move the method `onMessageWriteForPartition()` from `ReplicationEngine` to `ReplicationManager` (for better isolation). The component should be simple and lightweight, which is an inner class like `ClusterMapChangeListenerImpl` and `PartitionStateChangeListenerImpl`, what do you think?

Reply:
cc @cgtz  because you have a similar comment. Feel free to offer your insights here.